File: problems/拓扑排序/210.课程表II.cpp
#include <iostream>
#include <vector>
#include <queue>
using namespace std;

vector<int> findOrder(int numCourses, vector<vector<int>>& prerequisites) {
    vector<vector<int>> graph(numCourses);
    vector<int> indegree(numCourses, 0);
    for (auto& p : prerequisites) {
        graph[p[1]].push_back(p[0]);
        indegree[p[0]]++;
    }
    queue<int> q;
    for (int i = 0; i < numCourses; i++) 
        if (indegree[i] == 0) q.push(i);
    vector<int> order;
    while (!q.empty()) {
        int course = q.front(); q.pop();
        order.push_back(course);
        for (int neighbor : graph[course])
            if (--indegree[neighbor] == 0) q.push(neighbor);
    }
    return order.size() == numCourses ? order : vector<int>();
}

int main() {
    // 测试用例
    vector<vector<int>> p1 = {{1, 0}};
    auto r1 = findOrder(2, p1);
    for (int x : r1) cout << x << " ";
    cout << endl;  // 预期: 0 1
    
    vector<vector<int>> p2 = {{1, 0}, {2, 0}, {3, 1}, {3, 2}};
    auto r2 = findOrder(4, p2);
    for (int x : r2) cout << x << " ";
    cout << endl;  // 预期: 0 1 2 3
    
    vector<vector<int>> p3 = {};
    auto r3 = findOrder(1, p3);
    for (int x : r3) cout << x << " ";
    cout << endl;  // 预期: 0
    return 0;
}

File: problems/拓扑排序/210.课程表II.py
#!/usr/bin/env python3
"""
210. 课程表 II
返回完成所有课程的学习顺序。
"""
from collections import deque, defaultdict

def find_order(num_courses: int, prerequisites: list[list[int]]) -> list[int]:
    graph = defaultdict(list)
    indegree = [0] * num_courses
    for u, v in prerequisites:
        graph[v].append(u)
        indegree[u] += 1
    queue = deque([i for i in range(num_courses) if indegree[i] == 0])
    order = []
    while queue:
        course = queue.popleft()
        order.append(course)
        for neighbor in graph[course]:
            indegree[neighbor] -= 1
            if indegree[neighbor] == 0:
                queue.append(neighbor)
    return order if len(order) == num_courses else []

if __name__ == "__main__":
    # 测试用例
    print(find_order(2, [[1, 0]]))                    # 预期: [0, 1] 或其他有效顺序
    print(find_order(4, [[1, 0], [2, 0], [3, 1], [3, 2]]))  # 预期: [0, 1, 2, 3] 或其他有效顺序
    print(find_order(1, []))                         # 预期: [0]
    print(find_order(2, [[1, 0], [0, 1]]))            # 预期: [] (有环)
